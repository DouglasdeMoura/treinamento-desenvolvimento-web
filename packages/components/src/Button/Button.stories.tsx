import React from 'react';
import type { Meta } from '@storybook/react';

import { Button } from './Button';

export default {
  title: 'Button',
  component: Button,
} as Meta;

const Template = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  children: 'Label',
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
  children: 'Label',
};
