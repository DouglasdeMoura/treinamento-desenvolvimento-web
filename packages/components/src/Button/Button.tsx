import React, { forwardRef } from 'react';
import styled from 'styled-components';

type ButtonProps = {
  color?: string,
  background?: string,
  primary?: boolean,
} & JSX.IntrinsicElements['button'];

const ButtonComponent = forwardRef<HTMLButtonElement, ButtonProps>(({
  children,
  type = 'button',
  primary = false,
  ...rest
}, ref) => (
  <button type={type} data-primary={primary} {...rest} ref={ref}>
    {children}
  </button>
));

export const Button = styled(ButtonComponent)`
  border: 2px solid currentColor;
  border-radius: 4px;
  color: ${props => props.color ?? 'var(--color-red-500)'};
  cursor: pointer;
  background: ${props => props.background ?? 'white'};
  font-weight: bold;
  padding: 0.5em 1em;
  transition: filter 0.2s;

  :disabled {
    cursor: not-allowed;
    color: var(--color-gray-300);
  }

  &:focus {
    filter: brightness(120%);
  }

  &:not(:disabled):hover {
    filter: drop-shadow(0 0 0.75rem currentColor);
  }
`;
