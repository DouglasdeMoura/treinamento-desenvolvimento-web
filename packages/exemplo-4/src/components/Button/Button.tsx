import React, { forwardRef } from 'react';
import styled from 'styled-components';
import type { FC, MouseEvent, ReactNode } from 'react';

export type ButtonProps = {
  children: ReactNode,
  primary: boolean,
  backgroundColor: string,
  onClick(event?: MouseEvent<HTMLButtonElement>): void,
  size: 'small' | 'medium' | 'large',
};

const StyledButton = styled.button<Pick<ButtonProps, 'backgroundColor' | 'primary'>>`
  background: ${({ backgroundColor, primary }) =>
    backgroundColor ? backgroundColor : (primary  ? 'rebeccapurple' : '#f6f6f6')
  };
  border: none;
  border-radius: 50em;
  color: ${({ primary }) => primary ? 'white' : '#333'};
  font-weight: bold;

  &[data-size="small"] {
    font-size: 0.8rem;
    padding: 0.3rem 0.8rem;
  }

  &[data-size="medium"] {
    font-size: 1rem;
    padding: 0.5rem 1rem;
  }

  &[data-size="large"] {
    font-size: 1.2rem;
    padding: 0.7rem 1.4rem;
  }
`;

export const Button = forwardRef<HTMLButtonElement, FC<ButtonProps>>(({
  backgroundColor,
  children,
  onClick,
  primary = false,
  size = 'medium',
}, ref) => (
  <StyledButton
    data-size={size}
    backgroundColor={backgroundColor}
    primary={primary}
    onClick={onClick}
    ref={ref}
  >
    {children}
  </StyledButton>
));
