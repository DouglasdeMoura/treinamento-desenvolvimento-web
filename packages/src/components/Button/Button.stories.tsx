import React from 'react';
import { Button } from './Button';

import type { Meta, Story } from '@storybook/react';
import type { ButtonProps } from './Button';

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Normal = Template.bind({});
Normal.args = {
  children: 'Teste',
  primary: false,
  size: 'large',
};
