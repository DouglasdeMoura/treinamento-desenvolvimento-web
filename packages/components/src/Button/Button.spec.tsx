import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer';
import { Button } from './Button';

describe('<Button />', () => {
  it('deve renderizar o botão', () => {
    const Component = <Button>Nome</Button>;
    render(Component);
    const botao = screen.getByText('Nome');
    expect(botao).toHaveAttribute('data-primary', 'false');
    
    const tree = renderer
      .create(Component)
      .toJSON();
    
      expect(tree).toMatchSnapshot();
  });

  it('deve renderizar o botão primário', () => {
    render(<Button primary={true}>Nome</Button>);
    const botao = screen.getByText('Nome');
    expect(botao).toHaveAttribute('data-primary', 'true');
  });

  it('deve renderizar o nome do botão', () => {
    render(<Button>Nome</Button>);
    expect(screen.getByText('Nome')).toBeInTheDocument();
  });

  it('deve chamar função de OnClick', () => {
    const handleOnClick = jest.fn();
    render(<Button onClick={handleOnClick}>Nome</Button>);
    userEvent.click(screen.getByText('Nome'));
    expect(handleOnClick).toHaveBeenCalled();
  });
});
