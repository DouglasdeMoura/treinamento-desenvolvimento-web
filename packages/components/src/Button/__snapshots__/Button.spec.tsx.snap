// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> deve renderizar o bot√£o 1`] = `
<button
  className="sc-bdfBQB kvnXKx"
  data-primary={false}
  type="button"
>
  Nome
</button>
`;
